### ISSUE TITLE
Displaying outside temperture in touch bar

### ISSUE THREAD

**Issue by mark248am** [fifafu/BetterTouchTool#1696]
How can I display the outside temperature in the touch bar? Just something simple like "30°C"

thanks!

**Comment by fl0yd** [comment]
Touchbar settings -> +Widget -> Select Widget -> Run script and show return value.

You'll need to have a script that returns the value you expect before hand, but just point the widget at that.

**Comment by proactivematter** [comment]
This is a great idea. I would love to have this. I tried searching online for such as script but couldn't find any (this is my first mac - I just got it - I'm not familiar with these things).

**Comment by r3donnx** [comment]
Well, that is quite easy to achieve.
Open the terminal, and create a new weather.sh file:
`nano weather.sh`

Paste in the script below (This script is not mine, it is by Kaí Murciélago Frugívoro from askubuntu).
```
#!/bin/sh
w_xml=$(curl --silent "http://weather.tuxnet24.de/?id=20066057&mode=xml");
w_txt=$(xmllint --xpath "string(//current_text)" - <<<"$w_xml" | xargs);
w_tpc=$(xmllint --xpath "string(//current_temp)"  - <<<"$w_xml" | xargs); w_tpc=${w_tpc//[[:blank:]]/};

if   [ "$w_txt" == "Sunny" ]; then w_sym="☼";
elif [ "$w_txt" == "Mostly Sunny" ]; then w_sym="☼";
elif [ "$w_txt" == "Showers" ]; then w_sym="☂";
elif [ "$w_txt" == "Clear" ]; then w_sym="☾";
elif [ "$w_txt" == "Thunderstorms" ]; then w_sym="⚡";
elif [ "$w_txt" == "Scattered Thunderstorms" ]; then w_sym="☔";
elif [ "$w_txt" == "Isolated Thundershovers" ]; then w_sym="☔";
elif [ "$w_txt" == "Cloudy" ]; then w_sym="☁";
elif [ "$w_txt" == "Mostly Cloudy" ]; then w_sym="☁";
elif [ "$w_txt" == "Partly Cloudy" ]; then w_sym="☼☁";
elif [ "$w_txt" == "Breezy" ]; then w_sym="⚐";
else w_sym=$w_txt; 
fi
echo "$w_sym"" ""$w_tpc";
```
Now use your browser and navigate to the yahoo weather web page, change the location to the desired one and copy the location id from the address bar url.
Now paste in this location id to the second line of the script (20066057 should be changed)

Save the file by pressing CTRL+X, followed by "y" and ENTER

make this file executable:
`chmod +x weather.sh`

Good. now back to the BetterTouchTool.

Add a new touch bar widget. Select type "Run apple script and show return value".
Click the "Advanced configuration button"

Name the widget. "Weather" should be ok.

Paste in the code:
`return do shell script "/Users/YourUserName/weather.sh" `

Make sure to change "YourUserName" to your actual user name on the Mac. Or if you saved weather.sh to a different location, write the path to the weather.sh file.

Now click the "Run script" button to test if it is working.

**And make sure to change** "Execute this script every x seconds" value to 1800 or 3600 seconds or even more, so you are not flooding the weather server with the requests. I am updating results every half an our, that is 1800 seconds.

Hope it works.

**Comment by r3donnx** [comment]
You are welcome.
Well, it snows where I live too, but there was no snow this year yet :)

I guess when it starts to snow we can open the url from the script in the browser and take a look at the following condition:
`<current_text></current_text>`

I guess it will be something like
`<current_text>Snowing</current_text>` 

after that just add one more line to the script:

`elif [ "$w_txt" == "Snowing" ]; then w_sym="❆";`

Or any other icon you like ❄❅☃

**Comment by r3donnx** [comment]
I've rewritten the manual so now it uses AppleScript only, without the need to create the sh script in terminal. That makes it easier to setup.

So here is the new manual.
<img width="1085" alt="touch bar shot 2017-11-08 at 15 39 45" src="https://user-images.githubusercontent.com/23448062/32552014-33e14278-c49b-11e7-807f-6821d8384b1d.png">


Add a new touch bar widget of type "Run apple script and show return value".
Click the "Advanced configuration button"

Name the widget. "Weather" should be ok.

Paste in the code:

```
set locationId to "479616"

set weatherUrl to "http://weather.tuxnet24.de/?id=" & locationId & "&mode=xml"
set xmlData to text of (do shell script "curl " & quoted form of weatherUrl)

tell application "System Events"
	set rootXML to make new XML data with properties {text:xmlData}
	tell (XML element "info" of rootXML)
		set myTemp to (value of XML element "current_temp")
		set myText to (value of XML element "current_text")
	end tell
end tell
if myText is equal to "Sunny" then
	set myIcon to "☼"
else if myText is equal to "Mostly Sunny" then
	set myIcon to "☼"
else if myText is equal to "Showers" then
	set myIcon to "☂"
else if myText is equal to "Clear" then
	set myIcon to "☾"
else if myText is equal to "Cloudy" then
	set myIcon to "☁"
else if myText is equal to "Mostly Cloudy" then
	set myIcon to "☁"
else if myText is equal to "Partly Cloudy" then
	set myIcon to "☼☁"
else
	set myIcon to ""
end if
return myIcon & " " & myTemp
```

Use your browser and navigate to the yahoo weather web page, change the location to the desired one and copy the location id from the address bar url.
Now paste in this location id in to the first line of the script (479616 should be changed)

Now click the "Run script" button to test if it is working.

**And make sure to change "Execute this script every x seconds" value to 1800 or 3600 seconds or even more**, so you are not flooding the weather server with the requests. I am updating results every half an hour, that is 1800 seconds.

**Comment by proactivematter** [comment]
Awesome! Even better now. =)

I rearranged the conditions and added some emotes per my liking. Sharing here in case anyone else is interested.

I left "Snow"in there already so that when the time come I can just do a simple edit, but am of course not certain that "Snow" is the correct value.

```
set locationId to “0000”

set weatherUrl to "http://weather.tuxnet24.de/?id=" & locationId & "&mode=xml"
set xmlData to text of (do shell script "curl " & quoted form of weatherUrl)

tell application "System Events"
	set rootXML to make new XML data with properties {text:xmlData}
	tell (XML element "info" of rootXML)
		set myTemp to (value of XML element "current_temp")
		set myText to (value of XML element "current_text")
	end tell
end tell
if myText is equal to "Clear" then
	set myIcon to "🌈"
else if myText is equal to "Sunny" then
	set myIcon to "☀️"
else if myText is equal to "Mostly Sunny" then
	set myIcon to "🌤"
else if myText is equal to "Partly Cloudy" then
	set myIcon to "⛅️"
else if myText is equal to "Mostly Cloudy" then
	set myIcon to "🌥"
else if myText is equal to "Cloudy" then
	set myIcon to "☁️"
else if myText is equal to "Showers" then
	set myIcon to "🌧"
else if myText is equal to "Thunderstorms" then
	set myIcon to "⛈"
else if myText is equal to "Scattered Thunderstorms" then
	set myIcon to "🌩"
else if myText is equal to "Breezy" then
	set myIcon to "🌀"
else if myText is equal to "Snow" then
	set myIcon to "❄️"
else
	set myIcon to ""
end if
return myIcon & " " & myTemp
```

**Comment by fifafu** [comment]
Very nice scripts, I'll add this post to the knowledge base.

**Comment by robgraeber** [comment]
Looks like weather.tuxnet24.de got discontinued. 🙁

**Comment by proactivematter** [comment]
Yes, most unfortunately! Our enjoyment did not last long! I wonder if there are alternatives...
Best thing would be for this to run independently on our machines, pulling the information directly from the weather services, but I don't know how hard that would be to do.

**Comment by robgraeber** [comment]
I remade it using dark sky's api, they have free accounts up to 1000 req/day:

```
tell application "JSON Helper"
	set weather to fetch JSON from "https://api.darksky.net/forecast/[API_KEY]/37.8267,-122.4233?exclude=minutely,hourly,daily,alerts,flags"
	set temperature to temperature of currently of weather
	set weatherStatus to icon of currently of weather
end tell

if weatherStatus is equal to "clear-day" then
	set myIcon to "☀️"
else if weatherStatus is equal to "clear-night" then
	set myIcon to "🌙"
else if weatherStatus is equal to "partly-cloudy-day" or weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "⛅️"
else if weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "☁️🌙"
else if weatherStatus is equal to "cloudy" then
	set myIcon to "☁️"
else if weatherStatus is equal to "rain" then
	set myIcon to "🌧"
else if weatherStatus is equal to "sleet" then
	set myIcon to "⛈"
else if weatherStatus is equal to "fog" then
	set myIcon to "🌫️"
else if weatherStatus is equal to "wind" then
	set myIcon to "🌀"
else if weatherStatus is equal to "snow" then
	set myIcon to "❄️"
else
	set myIcon to weatherStatus
end if

return myIcon & " " & (round (temperature)) & "°"
```

**Comment by r3donnx** [comment]
weather.tuxnet24.de is online again. So my script should work as before

**Comment by robgraeber** [comment]
@proactivematter try renaming "JSON Helper" to "JSONHelper", BTT should prompt you to find the JSON Helper app and it should work after that. BTT has an issue finding JSON Helper for some reason.

**Comment by proactivematter** [comment]
Heya @robgraeber. I think it was my fault before, I must have missed some part of the code or copied something extra that first time, as I did manage to make it work without any modifications later.

I'm currently keeping both yours and r3donnx solutions so I can compare overtime and see which one is more accurate. I like DarkSky as it seems more precise and I am able to add Windchill ("feels like" temperature), but don't like it having fewer weather "types" (snow, rain, etc) in comparison to what Tuxnet pulls in from Yahoo.

Also, I've managed to make brew install locateme! So cool! I wonder what else I could use it for...

**Comment by tharris14** [comment]
Is there a way to make the temperature display in Fahrenheit?

**Comment by robgraeber** [comment]
@tharris14 it returns Fahrenheit for me but should be easy to add the conversion calc to the bottom

**Comment by tharris14** [comment]
Ok i have fixed that, but when it runs, now it says Error, "Can't get currently of "" ", for the line:
"set temperature to temperature of currently of weather"

**Comment by r3donnx** [comment]
If you use the script I offered, just add `&unit=imperial` to the end of the weatherUrl variable.
Like this:
`set weatherUrl to "http://weather.tuxnet24.de/?id=" & locationId & "&mode=xml&unit=imperial"`

**Comment by draculaentsteint** [comment]
since a few days i'm always getting an error message when using the applescript (the one from proactivematter to be exact)

> System Events got an error: Can’t get XML element "info" of XML data id "untitled 3".

anyone having the same issue?

**Comment by KalleJanssen** [comment]
Yeah I have the same issue.

**Comment by OsosKP** [comment]
I don't know if anybody else is experiencing this, but I was having two issues that caused this script to stop working:

1. The given XML had been moved to another location
2. The data in the XML itself had been changed to plaintext, and so XMLLINT wasn't working

I rewrote the script to process the plaintext and give the same result as above:

#!/bin/sh
w_xml=$(curl --silent "https://weather.tuxnet24.de/?id=560743&amp;mode=xml");
# w_txt=$(xmllint --xpath "string(//current_text)" - <<<"$w_xml" | xargs);
# awk '$1 == "current_text" {w_txt=$3}' - <<< "$w_xml" | xargs
# "$w_xml" > awk '/current_text/ {echo $3}'
line=$(echo "$w_xml" | grep current_text)
words=( $line )
w_txt=${words[2]}
# w_tpc=$(xmllint --xpath "string(//current_temp)"  - <<<"$w_xml" | xargs); w_tpc=${w_tpc//[[:blank:]]/};
# awk '/current_temp/ {w_tpc=$3}' - <<< "$w_xml" | xargs
line=$(echo "$w_xml" | grep current_temp)
words=( $line )
w_tpc=${words[2]}

if   [ "$w_txt" == "Sunny" ]; then w_sym="☼";
elif [ "$w_txt" == "Mostly Sunny" ]; then w_sym="☼";
elif [ "$w_txt" == "Showers" ]; then w_sym="☂";
elif [ "$w_txt" == "Clear" ]; then w_sym="☾";
elif [ "$w_txt" == "Thunderstorms" ]; then w_sym="⚡";
elif [ "$w_txt" == "Scattered Thunderstorms" ]; then w_sym="☔";
elif [ "$w_txt" == "Isolated Thundershovers" ]; then w_sym="☔";
elif [ "$w_txt" == "Cloudy" ]; then w_sym="☁";
elif [ "$w_txt" == "Mostly Cloudy" ]; then w_sym="☁";
elif [ "$w_txt" == "Partly Cloudy" ]; then w_sym="☁";
elif [ "$w_txt" == "Breezy" ]; then w_sym="⚐";
else w_sym=$w_txt; 
fi
echo "$w_sym"" ""$w_tpc";


Note: I had previously changed the icons used so be sure to change them back if you want.

**Comment by robgraeber** [comment]
The weather script I posted still works great 😛

**Comment by KalleJanssen** [comment]
Thanks @OsosKP it works!

**Comment by robgraeber** [comment]
If anyone prefers a python version of the dark sky script:

`pip install requests`
```python
import os
import requests
latLon = os.popen("/usr/local/bin/locateme -f '{LAT},{LON}'").readlines()[0].rstrip()
apiUrl = 'https://api.darksky.net/forecast/API_KEY_HERE/' + latLon + '?exclude=minutely,hourly,daily,alerts,flags'
rsp = requests.get(apiUrl).json()
temperature = rsp['currently']['temperature']
weatherStatus = rsp['currently']['icon']

iconDict = {
  'clear-day': "☀️",
  'clear-night': "☀️",
  'partly-cloudy-day': "⛅️",
  'partly-cloudy-night': "⛅️",
  'cloudy': "☁️",
  'rain': "🌧",
  'sleet': "⛈",
  'fog': "🌫️",
  'wind': "🌀",
  'snow': "❄️"
}

print iconDict[weatherStatus] + " " + str(round(temperature * 10) / 10) + "°"
```

**Comment by tatum07** [comment]
@robgraeber I'm using your script, but it is showing the temperature in fahrenheight. I am from Canada, as such I'd love to have it displayed in Celsius. I am not very good at programming, can you tell me what to add, and where to add it to display in Celsius.

**Comment by fifafu** [comment]
(T(°F) - 32) × 5/9
I think you can just replace the line
```
temperature = rsp['currently']['temperature']
```

with

```
temperature = (rsp['currently']['temperature'] - 32) * (5/9)
```

**Comment by tatum07** [comment]
@fifafu  thanks for your response, but I'm using the other script:

tell application "JSON Helper"
	set latLon to do shell script "/usr/local/bin/locateme -f '{LAT},{LON}'"
	set weather to fetch JSON from ("https://api.darksky.net/forecast/[API_KEY]/" & latLon & "?exclude=minutely,hourly,daily,alerts,flags")
	set temperature to temperature of currently of weather
	set weatherStatus to icon of currently of weather
end tell

if weatherStatus is equal to "clear-day" then
	set myIcon to "☀️"
else if weatherStatus is equal to "clear-night" then
	set myIcon to "🌙"
else if weatherStatus is equal to "partly-cloudy-day" or weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "⛅️"
else if weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "☁️🌙"
else if weatherStatus is equal to "cloudy" then
	set myIcon to "☁️"
else if weatherStatus is equal to "rain" then
	set myIcon to "🌧"
else if weatherStatus is equal to "sleet" then
	set myIcon to "⛈"
else if weatherStatus is equal to "fog" then
	set myIcon to "🌫️"
else if weatherStatus is equal to "wind" then
	set myIcon to "🌀"
else if weatherStatus is equal to "snow" then
	set myIcon to "❄️"
else
	set myIcon to weatherStatus
end if

return myIcon & (round (temperature)) & "°"


Any idea where to add here?

**Comment by fifafu** [comment]
While I really recommend to use the python version (much more performant) you can e.g.  just put it in the return at the end. Just replace temperature with (temperature-32)*(5/9)

**Comment by FinanceTutor** [comment]
Let's rather put it this way: I think my problem is, that I don't really know how to install locateme /where to download it... Is there some download link..? Please excuse my lack of knowledge and thank's for the workaround already!

**Comment by robgraeber** [comment]
Brew is a package manager for mac, see https://brew.sh

TLDR:
```
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
brew install locateme
```

**Comment by robgraeber** [comment]
Did you sign up for a darksky api key?

**Comment by kdumovic** [comment]
@robgraeber Add this line to the top of your Python file so it doesn't complain about the non-ASCII characters (emojis).

`# -*- coding: utf-8 -*-`

**Comment by robgraeber** [comment]
@kdumovic Hmm I didn't have that issue, but sure added it.

**Comment by iconio** [comment]
@robgraeber Hey how's it going? 
I'm trying to use the script below but I'm getting an error when I compile, it says "Syntax Error: Expected end of line but found identifier".

I have also created the API Key, and I am setting my location as well (LAT, LON). I also download from the Mac AppStore the JSON Helper.

What am I missing?

Thanks bud
============
tell application "JSON Helper"
	set weather to fetch JSON from "https://api.darksky.net/forecast/[API_KEY]/37.8267,-122.4233?exclude=minutely,hourly,daily,alerts,flags"
	set temperature to temperature of currently of weather
	set weatherStatus to icon of currently of weather
end tell

if weatherStatus is equal to "clear-day" then
	set myIcon to "☀️"
else if weatherStatus is equal to "clear-night" then
	set myIcon to "🌙"
else if weatherStatus is equal to "partly-cloudy-day" or weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "⛅️"
else if weatherStatus is equal to "partly-cloudy-night" then
	set myIcon to "☁️🌙"
else if weatherStatus is equal to "cloudy" then
	set myIcon to "☁️"
else if weatherStatus is equal to "rain" then
	set myIcon to "🌧"
else if weatherStatus is equal to "sleet" then
	set myIcon to "⛈"
else if weatherStatus is equal to "fog" then
	set myIcon to "🌫️"
else if weatherStatus is equal to "wind" then
	set myIcon to "🌀"
else if weatherStatus is equal to "snow" then
	set myIcon to "❄️"
else
	set myIcon to weatherStatus
end if

return myIcon & " " & (round (temperature)) & "°"

**Comment by lhawro** [comment]
> While I really recommend to use the python version (much more performant) you can e.g. just put it in the return at the end. Just replace temperature with (temperature-32)*(5/9)

With Python script this Celsius fix didn't work for me - the temperature was showing 0.0 for some reason
I've adjusted the last line of the code 
[print iconDict[weatherStatus] + " " + str(round(((temperature - 32) / 1.8) * 10) / 10) + "°"](url)
